### PR TITLE
[stable10] tests to create users by supplying and email address an no password

### PR DESCRIPTION
**Pull request by individual-it**
## Description
UI tests for "New User Registration Page" and the workflow where the administrator can create a user without providing a password. The user will be notified by Email and can set the password himself

## Related Issue
-  #32560

## How Has This Been Tested?
:robot: 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [x] Tests

## Checklist:

- [ ] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 

## Open tasks:

- [ ] make a similar PR to user_management app

### CODE REVIEW

File: tests/TestHelpers/EmailHelper.php [hunk 1]
```diff
@@ -59,9 +59,29 @@ public static function getEmails($mailhogUrl) {
...
+	 *
+	 * @param string $mailhogUrl
+	 * @param string $address
+	 * @param int $no which number of multiple emails to read (first email is 1)
+	 *
+	 * @throws \Exception
+	 *
+	 * @return mixed
+	 */
+	public static function getBodyOfEmail($mailhogUrl, $address, $no=1) {
```
Comment (phil-davis) [inline tests/TestHelpers/EmailHelper.php]: we usually put spaces like:
```
$no = 1
```
If you change this, then might as well make the var name more descriptive ``$numEmails`` ?

File: tests/TestHelpers/EmailHelper.php [hunk 2]
```diff
@@ -59,9 +59,29 @@ public static function getEmails($mailhogUrl) {
...
+		$skip=1;
```
Comment (phil-davis) [inline tests/TestHelpers/EmailHelper.php]:
```
$skip = 1;
```

File: tests/acceptance/features/bootstrap/WebUIGeneralContext.php [hunk 3]
```diff
@@ -219,6 +231,23 @@ public function theUserLogsOutOfTheWebUI() {
...
 	}
 
+	/**
+	 *
+	 * @param string $emailAddress
+	 * @param string $regexSearch
+	 * @param string $errorMessage which number of multiple emails to read (first email is 1)
+	 * @param int $no
+	 */
+	public function followLinkFromEmail($emailAddress, $regexSearch, $errorMessage, $no=1) {
```
Comment (phil-davis) [inline tests/acceptance/features/bootstrap/WebUIGeneralContext.php]:
```
$no = 1
```
and makeup a longer var name

File: tests/acceptance/features/bootstrap/WebUIGeneralContext.php [hunk 4]
```diff
@@ -367,6 +396,33 @@ public function theUserShouldBeRedirectedToAWebUIPageWithTheTitle($title) {
 		PHPUnit_Framework_Assert::assertEquals($title, $this->owncloudPage->getPageTitle());
 	}
 
+	/**
+	 * @Then the user should be redirected to general error webUI page with the title :title
```
Comment (phil-davis) [inline tests/acceptance/features/bootstrap/WebUIGeneralContext.php]: to the general error webUI page

File: tests/acceptance/features/bootstrap/WebUILoginContext.php [hunk 5]
```diff
@@ -294,6 +287,24 @@ public function theUserResetsThePasswordToUsingTheWebui($newPassword) {
 		$this->loginPage->resetThePassword($newPassword, $this->getSession());
 	}
 
+	/**
+	 * @When /^the user follows the password set link received by "([^"]*)"(?: in Email no (\d+))? using the webUI$/
```
Comment (phil-davis) [inline tests/acceptance/features/bootstrap/WebUILoginContext.php]: s/no/number

File: tests/acceptance/features/bootstrap/WebUILoginContext.php [hunk 6]
```diff
@@ -294,6 +287,24 @@ public function theUserResetsThePasswordToUsingTheWebui($newPassword) {
...
+	 *
+	 * @param string $emailAddress
+	 * @param int which number of multiple emails to read (first email is 1)
```
Comment (phil-davis) [inline tests/acceptance/features/bootstrap/WebUILoginContext.php]: PHPdoc needs the var name

File: tests/acceptance/features/bootstrap/WebUILoginContext.php [hunk 7]
```diff
@@ -294,6 +287,24 @@ public function theUserResetsThePasswordToUsingTheWebui($newPassword) {
...
+	 *
+	 * @return void
+	 * @throws \Exception
+	 */
+	public function theUserFollowsThePasswordSetLinkReceivedByEmail($emailAddress, $no=1) {
```
Comment (phil-davis) [inline tests/acceptance/features/bootstrap/WebUILoginContext.php]:
```
$var = 1
```
spacing

File: tests/acceptance/features/webUIManageUsersGroups/addUsers.feature [hunk 8]
```diff
@@ -47,3 +47,63 @@ Feature: add users
...
+    When the administrator creates a user with the name "guiusr1" and the email "guiusr1@owncloud" without a password using the webUI
+    Then the email address "guiusr1@owncloud" should have received an email with the body containing
+      """
+      just letting you know that you now have an ownCloud account.
+      
+      Your username: guiusr1
+      Access it:
+      """
+
+  Scenario Outline: user sets his own password after beeing created with an Email address only
```
Comment (phil-davis) [inline tests/acceptance/features/webUIManageUsersGroups/addUsers.feature]: s/beeing/being

File: tests/acceptance/features/webUIManageUsersGroups/addUsers.feature [hunk 9]
```diff
@@ -47,3 +47,63 @@ Feature: add users
...
+    And the user follows the password set link received by "guiusr1@owncloud" using the webUI
+    And the user sets the password to "newpassword" using the webUI
+    Then the email address "guiusr1@owncloud" should have received an email with the body containing
+      """
+      Password changed successfully
+      """
+    When the user logs in with username "<username>" and password "newpassword" using the webUI
+    Then the user should be redirected to a webUI page with the title "Files - ownCloud"
+    Examples:
+     | username | comment |
```
Comment (phil-davis) [inline tests/acceptance/features/webUIManageUsersGroups/addUsers.feature]: line up the columns

File: tests/acceptance/features/webUIManageUsersGroups/addUsers.feature [hunk 10]
```diff
@@ -47,3 +47,63 @@ Feature: add users
...
+      |Error creating user: Invalid mail address|
+    Examples:
+      |email   | comment       |
+      | string | no @ sign     |
+      | a@     | no domain name|
+      #there would be much more to test here, but its complicated and would be slow
+      #see http://codefool.tumblr.com/post/15288874550/list-of-valid-and-invalid-email-addresses
+      #email address validation would better go into an unit test
+
+  Scenario: webUI refuses to create a user with an empty Email addresse
```
Comment (phil-davis) [inline tests/acceptance/features/webUIManageUsersGroups/addUsers.feature]: s/addresse/address

File: tests/acceptance/features/webUIManageUsersGroups/addUsers.feature [hunk 11]
```diff
@@ -47,3 +47,63 @@ Feature: add users
...
+  Scenario: changing the user password as an admin invalidates the user sets-password-token
+    When the administrator creates a user with the name "guiusr1" and the email "guiusr1@owncloud" without a password using the webUI
+    And the administrator changes the password of user "guiusr1" to "123" using the provisioning API
+    And the administrator logs out of the webUI
+    And the user follows the password set link received by "guiusr1@owncloud" using the webUI
+    Then the user should be redirected to general error webUI page with the title "ownCloud"
+    And an error should be displayed on the general error webUI page saying "The token provided is invalid."
+
+  Scenario: sets-password-token cannot be used twice
+      When the administrator creates a user with the name "guiusr1" and the email "guiusr1@owncloud" without a password using the webUI
```
Comment (phil-davis) [inline tests/acceptance/features/webUIManageUsersGroups/addUsers.feature]: tab indenting here is different to the other scenarios

**Review by phil-davis — COMMENTED** [review]
It works - just rubbish review comments!

**Comment by individual-it** [comment]
addresses review comments, tests failing because of #32647

**Review by phil-davis — APPROVED** [review]

**Comment by phil-davis** [comment]
"backport-request" means forward-port splitting the relevant bits to core master and to the user_management app.

**Comment by individual-it** [comment]
forward port in https://github.com/owncloud/core/pull/32667
and https://github.com/owncloud/user_management/pull/65

**Pull request merged**